### ISSUE TITLE
Can this predict directed graphs?

### ISSUE THREAD

**Issue by aalex1945** [KrzakalaPaul/Any2Graph#1]
It is great work. However, matrix A is said to be symmetric in the paper. I wonder if this model can generate directed graphs. Or can it be relatively easily adapted to generate directed graphs? Thanks.

**Comment by KrzakalaPaul** [comment]
Hello, thank you for your message. 
Good question, yes it can be adapted, it just takes a few careful changes. Since you are not the first one to ask I just made the required update :)
You can now set a "symmetric" flag in the config, "symmetric = False" changes the pipeline in 2 ways:

- The decoder becomes asymmetric: A_{i,j} = f( [x_i,x_j] ) instead of A_{i,j} = f( x_i + x_j ) 
- The solver for computing the loss slightly changes. Some factorizations do not hold in the asymmetric case so its slightly more expensive

**Comment by aalex1945** [comment]
Thank you!!!!

**Comment by salcc** [comment]
Hi, thanks for this great work. I have a question regarding the implementation of the update. Why don't you use `Lt` in the cost and loss computations? For instance, something like:

```python
LT_f = tensor_product_quad_batch(L,  T, …)
LT_b = tensor_product_quad_batch(Lt, T, …)
LT   = 0.5*(LT_f + LT_b)
losses_A = torch.sum(T * LT)/B
```

and similarly in the cost computations.

**Comment by KrzakalaPaul** [comment]
Hi, good question. In theory it should hold that

torch.sum(T * LT_f) = torch.sum(T * LT_b) = torch.sum(T * LT)

(its not true if you compute torch.sum(M * LT) for some M != T)

Thus what you propose might be slightly more numerically stable but mathematically it is not required (and its an extra computation).

**Comment by KrzakalaPaul** [comment]
To be more clear the goal is to minimize F(T) = < T, L x T > where x is the tensor product. 

It turns out that L is a symmetric operator ( < A, L x B > = < L x A, B >) if and only if C1 and C2 are symmetric matrices. 
If they are not then we have to introduce Lt, the conjugate of L such that < A, Lt x B > = < Lt x A, B >.

In that case we have that:

grad F(T) = Lt x T + L x T

(in the symmetric case its simply grad F(T) = 2 L x T)

Therefore we have to compute Lt for the optimization process even though the target objective satisfies 

F(T) = < T, L x T > = < T, Lt x T > = 0.5 < T, (L+Lt) x T >
